text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+done just like they are with regular integer literals.
+
+Primitive operations on `int<N>` and `uint<N>` should work exactly like they do
+on `int<N>` and `uint<N>`: overflows should panic when debug assertions are
+enabled, but ignored when they are not. In general, `uint<N>` will be
+zero-extended to the next power of two, and `int<N>` will be sign-extended to
+the next power of two.
+
+## Computations and storage
+
+Because sign extension will always be applied, it's safe for the compiler to
+internally treat `uint<N>` as `uint<N.next_power_of_two()>` when doing all
+computations. As a concrete example, this means that adding two `uint<48>`
+values will work exactly like adding two `u64` values, generating exactly the
+same code on targets with 64-bit registers. For targets with 32-bit registers,
+it may generate slightly different code, although in practice it'll probably be
+a long time before these sorts of optimisations are applied.
+
+Additionally, the restriction that `N <= 128` may allow for future changes
+where `uint<N>` is not actually stored as `uint<N.next_power_of_two()>` for
+sufficiently large `N`. Right now, even storing `uint<65>` as `u128` makes the
+most sense, because it will be the most efficient representation on 64-bit CPUs
+for performing many operations. However, storing `uint<4097>` as `uint<8192>`,
+for example, may be total overkill.
+
+The main goal for these representations is to avoid bit-wrangling as much as
+possible. Although truncating the number of bits for integers is no longer free,
+e.g. `u64` to `uint<48>` requires an AND and `i64` to `int<48>` requires
+conditionally excecuting an AND or an OR, the goal is that in most cases, using
+non-power-of-two lengths should be more or less free.
+
+## Niche optimizations for enums
+
+As a hard rule, `uint<N>` will always have its upper `N.next_power_of_two() - N`
+bits set to zero, and similarly, `int<N>`'s upper bits will be sign-extended. To
+compute niche optimizations for enums, we'll have to take this into account.
+
+Niche values for `uint<N>` are simply any nonzero patterns in the upper bits,
+similarly to `bool`. This means that for `uint<7>`, for example, we have `128`
+possible niche values.
+
+Niche values for `int<N>` are trickier. The niche with all bits set to one
+represents the valid value -1, and should be replaced with a niche where all
+the `int<N>` bits are set to one and the extra bits are set to zero. Other than
+that, in general, both `uint<N>` and `int<N>` will have
+`N.next_power_of_two() - N` bits of niche space.
+
+## Standard library
+
+Existing implementations for integer types should be annotated with
+`default impl` as necessary, and most operations should defer to the
+implementations for `N.next_power_of_two()`. For example, the `count_zeros`
+function in the generic impl would be:
+
+```rust
+impl<const N: usize> uint<N> {
+    fn count_zeros(self) -> u32 {
+        let M = N.next_power_of_two();
```

Comment:
I would assume this has to be `const M = N.next_power_of_two();` since you can't use a non-const value in the type parameter on the next line. This [appears to not be allowed by RFC2000](https://github.com/rust-lang/rfcs/blob/master/text/2000-const-generics.md#when-a-const-variable-can-be-used) though.

It seems that this could be written

```rust
impl<const N: usize> uint<N> {
    fn count_zeros(self) -> u32 {
        let zeros = (self as uint<{ N.next_power_of_two() }>).count_zeros();
        zeros + (N.next_power_of_two() - N)
    }
}
```

but I'm not certain if the const expression there would be accepted by the current const generics implementation.

Reply:
You're right and I replaced `let M` with `const M: usize` for now. I think that's valid.